docs/nnx/filters_guide.md

```diff
@@ -0,0 +1,176 @@
+
+If you wish to manually create a callable from a literal to can use `nnx.filterlib.to_predicate`:
+
+```{code-cell}
+is_param = nnx.filterlib.to_predicate(nnx.Param)
+everything = nnx.filterlib.to_predicate(...)
+nothing = nnx.filterlib.to_predicate(False)
+any = nnx.filterlib.to_predicate([nnx.Param, 'dropout'])
+
+print(f'{type(is_param) = }')
+print(f'{type(everything) = }')
+print(f'{type(nothing) = }')
+print(f'{type(any) = }')
+```
+
+## Grouping States
+
+With the knowledge of Filters at hand, let's see how `nnx.split` is roughly implemented. Key ideas:
+
+* Use `nnx.graph.flatten` to get the `GraphDef` and `State` representation of the node.
+* Convert all the filters to predicates.
+* Use `State.flat_state` to get the flat representation of the state.
+* Traverse all the `(path, value)` pairs in the flat state and group them according to the filters.
+* Use `State.from_flat_state` to convert the flat states to nested `State`s.
+
+```{code-cell}
+from typing import Any
+KeyPath = tuple[nnx.graph.Key, ...]
+
+def split(node, *filters):
+  graphdef, state, _ = nnx.graph.flatten(node)
+  predicates = [nnx.filterlib.to_predicate(f) for f in filters]
+  flat_states: list[dict[KeyPath, Any]] = [{} for p in predicates]
+
+  for path, value in state.flat_state().items():
+    for i, predicate in enumerate(predicates):
+      if predicate(path, value):
+        flat_states[i][path] = value
+        break
+    else:
+      raise ValueError(f'No filter matched {path = } {value = }')
+    
+  states: tuple[nnx.GraphState, ...] = tuple(
+    nnx.State.from_flat_path(flat_state) for flat_state in flat_states
+  )
+  return graphdef, *states
+
+# lets test it...
+foo = Foo()
+
+graphdef, params, batch_stats = split(foo, nnx.Param, nnx.BatchStat)
+
+print(f'{params = }')
+print(f'{batch_stats = }')
+```
+
+One very important thing to note is that **the order of the filters is important**. The first filter that matches a value will be the group that the value is placed in, therefore you should place more specific filters before more general filters. For example if we create a `SpecialParam` type that is a subclass of `Param` and create an object `Bar` that contains both types of parameters, if we try to split the `Param`s before the `SpecialParam`s, all the values will be placed in the `Param` group and the `SpecialParam` group will be empty because all `SpecialParam`s are also `Param`s:
+
+```{code-cell}
+class SpecialParam(nnx.Param):
+  pass
+
+class Bar(nnx.Module):
+  def __init__(self):
+    self.a = nnx.Param(0)
+    self.b = SpecialParam(0)
+
+bar = Bar()
+
+graphdef, params, special_params = split(bar, nnx.Param, SpecialParam) # wrong!
+print(f'{params = }')
+print(f'{special_params = }')
+```
+
+Reversing the order will make sure that the `SpecialParam` are captured first
+
+```{code-cell}
+graphdef, special_params, params = split(bar, SpecialParam, nnx.Param) # correct!
```

Comment:
Since we introduced the filter DSL above, it would be cool to showcase it here in this example!

Reply:
Added an example using the DSL in `nnx.vmap`